Mathlib/RingTheory/HahnSeries.lean

```diff
@@ -618,22 +622,111 @@ theorem order_one [MulZeroOneClass R] : order (1 : HahnSeries Γ R) = 0 := by
   · exact order_single one_ne_zero
 #align hahn_series.order_one HahnSeries.order_one
 
-instance [NonUnitalNonAssocSemiring R] : Mul (HahnSeries Γ R) where
-  mul x y :=
-    { coeff := fun a =>
-        ∑ ij in addAntidiagonal x.isPWO_support y.isPWO_support a, x.coeff ij.fst * y.coeff ij.snd
-      isPWO_support' :=
+end HahnSeries
+
+/-- We introduce a type alias for `HahnSeries` in order to work with scalar multiplication by
+series. If we wrote a `SMul (HahnSeries Γ R) (HahnSeries Γ V)` instance, then when
+`V = HahnSeries Γ R`, we would have two different actions of `HahnSeries Γ R` on `HahnSeries Γ V`.
+See `Mathlib.Data.Polynomial.Module` for more discussion on this problem. -/
+@[nolint unusedArguments]
+def HahnModule (Γ R V : Type*) [PartialOrder Γ] [Zero V] [SMul R V] :=
+  HahnSeries Γ V
+
+namespace HahnModule
+
+/-- The casting function to the type synonym. -/
+def of {Γ : Type*} (R : Type*) {V : Type*} [PartialOrder Γ] [Zero V] [SMul R V] :
+    HahnSeries Γ V ≃ HahnModule Γ R V := Equiv.refl _
```

Comment:
Should this be marked reducible, because it's now used to define the `[NonUnitalNonAssocSemiring R] : Mul (HahnSeries Γ R)` instance?

Reply:
That sounds like it might be a good idea (but I don't know enough to judge).  I'll try putting it in #10846, which is still open.